### DISCUSSION TITLE
Optionally using user-provided Raw Socket/net.Conn for UDP communications

### DISCUSSION THREAD

**Discussion by gaukas** [pion/webrtc#1948] [Q&A] [answered]
## Summary

Allow user (caller) to provide the base level `net.Conn` /raw socket for UDP communications.

## Motivation

This feature would allow library users to have control at a much lower layer over the packets sent during the WebRTC communication, which would fit more network structure and enable many fancy use cases. 

## Describe alternatives you've considered

Or could you please provide clear instruction to the line of code where the raw socket (net.Conn) has been wrapped into the library so I could fork and try it on my own.

## Additional context

I have been working on an Internet Censorship Circumventing project called **Conjure** and our current network topology needs access to the raw socket in order to make it work.

[Paper CCS'19](https://dl.acm.org/doi/pdf/10.1145/3319535.3363218)
[Repository on GitHub](https://github.com/refraction-networking/conjure)

**Comment by gaukas** [comment]
I'm sorry if I sounded like someone didn't read the code and come to demand... I have spent several hours jumping between all those dependencies trying to find the correct `net.Conn` struct. 😥

**Comment by Sean-Der** [comment, marked as answer]
Very cool project @Gaukas! Excited you are using Pion :)

Yes this is possible today! Via the SettingEngine you can use the [ICEUDPMux](https://pkg.go.dev/github.com/pion/webrtc/v3#SettingEngine.SetICEUDPMux). This allows you to pass in a socket you are already listening and the PeerConnection uses that.
